### PR TITLE
Feature pooled objects

### PR DESCRIPTION
**Pull request by vijit101**
-Added Pool and corrected namespaces working well

### CODE REVIEW

File: Assets/Scripts/Bullet/BulletService.cs [hunk 1]
```diff
@@ -23,18 +23,21 @@ public BulletController GetBullet(TankType tankType)
             {
                 BulletModel bulletModel = new BulletModel(bulletScriptableObjectList.Bullets[0]);
                 BulletController bulletController = bulletControllerPoolService.GetBulletController(bulletModel, bulletView);
+                bulletController.BulletView.gameObject.SetActive(true);
```
Comment (mayankgrover) [inline Assets/Scripts/Bullet/BulletService.cs:26]: bullet view / game object should not be visible outside the bullet controller. 
Abstraction - you need to hide away the details internal to the controller

File: Assets/Scripts/Enemy/EnemyBehaviour.cs [hunk 2]
```diff
@@ -47,7 +48,9 @@ public void TakeDamage(float Damage)
                 score++;
                 PlayerPrefs.SetFloat("Score", score);
                 //Enemy Death state
-                Destroy(gameObject);
+                //return the enemy to the pool
+                this.gameObject.SetActive(false);
+                EnemyPoolService.Instance.ReturnPooledObject(this.gameObject);
```
Comment (mayankgrover) [inline Assets/Scripts/Enemy/EnemyBehaviour.cs:53]: return a specific Controller/view to the pool not the game object

Comment (mayankgrover) [reply Assets/Scripts/Enemy/EnemyBehaviour.cs:53]: the pool should be of the controller or the view - never just game objects

File: Assets/Scripts/Enemy/EnemyService.cs [hunk 3]
```diff
@@ -43,8 +46,12 @@ public void SpawnEnemyTank()
             enemySpawnPos = enemySpawnPos.SetRandomVectorXYZ(-39, 39, 0, 10, -39, 39);
             enemySpawnPos = enemySpawnPos.SetY(-3.68f);
             EnemyTankPositions.Add(enemySpawnPos);
-            GameObject enemyTank = Instantiate(EnemyTankViews[Random.Range(0, EnemyTankViews.Count)], enemySpawnPos, Quaternion.identity) as GameObject;
-            InstantiatedEnemyTanks.Add(enemyTank);
+            GameObject enemyTank = enemyPoolService.GetEnemyTank(EnemyTankViews[Random.Range(0, EnemyTankViews.Count)]);
```
Comment (mayankgrover) [inline Assets/Scripts/Enemy/EnemyService.cs:49]: same thing as above - pool should be of a specific class (controller or view) not gameobject

File: Assets/Scripts/Enemy/EnemyService.cs [hunk 4]
```diff
@@ -7,15 +8,17 @@ public class EnemyService : MonoSingletongeneric<EnemyService>
     {
         public List<GameObject> EnemyTankViews;
         public List<Vector3> EnemyTankPositions;
-        public List<GameObject> InstantiatedEnemyTanks;
+        EnemyPoolService enemyPoolService;
```
Comment (mayankgrover) [inline Assets/Scripts/Enemy/EnemyService.cs:11]: service should be a singleton

File: Assets/Scripts/Enemy/EnemyService.cs [hunk 5]
```diff
@@ -43,8 +46,12 @@ public void SpawnEnemyTank()
             enemySpawnPos = enemySpawnPos.SetRandomVectorXYZ(-39, 39, 0, 10, -39, 39);
             enemySpawnPos = enemySpawnPos.SetY(-3.68f);
             EnemyTankPositions.Add(enemySpawnPos);
-            GameObject enemyTank = Instantiate(EnemyTankViews[Random.Range(0, EnemyTankViews.Count)], enemySpawnPos, Quaternion.identity) as GameObject;
-            InstantiatedEnemyTanks.Add(enemyTank);
+            GameObject enemyTank = enemyPoolService.GetEnemyTank(EnemyTankViews[Random.Range(0, EnemyTankViews.Count)]);
+            enemyTank.transform.position = enemySpawnPos;
+            enemyTank.transform.rotation = Quaternion.identity;
+            enemyTank.SetActive(true);
```
Comment (mayankgrover) [inline Assets/Scripts/Enemy/EnemyService.cs:52]: setting position/rotation/enable/disable should be functions on the controller and not done in the service level

**Review by mayankgrover — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Enemy/EnemyService.cs (hunk 4 above)
Comment (vijit101) [reply Assets/Scripts/Enemy/EnemyService.cs:11]: Solved by using Instance of pool service

File: Assets/Scripts/Bullet/BulletService.cs (hunk 1 above)
Comment (vijit101) [reply Assets/Scripts/Bullet/BulletService.cs:26]: Abstracted in the bullet pooler service itself as if getBulletItem is not null set active true

File: Assets/Scripts/Enemy/EnemyService.cs (hunk 5 above)
Comment (vijit101) [reply Assets/Scripts/Enemy/EnemyService.cs:52]: No Controller made for enemy behavior kept it simple but did used a method to reset enemy too  and  I used the review and corrected bullet and player logic as added  reset position() logic  to the controller

File: Assets/Scripts/Enemy/EnemyService.cs (hunk 3 above)
Comment (vijit101) [reply Assets/Scripts/Enemy/EnemyService.cs:49]: Done Created A Pool of my EnemyBehaviour script ie. the view script for enemy

File: Assets/Scripts/Enemy/EnemyBehaviour.cs (hunk 2 above)
Comment (vijit101) [reply Assets/Scripts/Enemy/EnemyBehaviour.cs:53]: Done added view

**Pull request merged**